File: lab4/A.py
from sys import stdin, stdout
class Node:
    def __init__(self, pnt):
        self.data = pnt
        self.r = None
        self.l = None

def insert(root, pnt):
    if root is None:
        return Node(pnt)
    else:
        if root.data == pnt:
            return root
        elif root.data < pnt:
            root.r = insert(root.r, pnt)
        else:
            root.l = insert(root.l, pnt)
    return root

def max_depth(node):
    if node is None:
        return 0
    else:
        left_depth = max_depth(node.l)
        right_depth = max_depth(node.r)

        if (left_depth > right_depth):
            return left_depth+1
        else:
            return right_depth+1
mp = map
lst = list
arr = lst(mp(int, stdin.readline().strip().split()))
parent = Node(arr[0])
for i in arr[1:-1]:
    insert(parent,i)
stdout.write(str(max_depth(parent)))

File: lab4/A.cpp
#include <iostream>

using namespace std;

struct Node{
    int data;
    Node *left, *right;
    Node(int x):left(NULL), right(NULL),data(x){}
};

class BST{
    private:
    Node *root;
    void _print(Node *node){
        if (node == NULL)
            return;
        _print(node->left);
        cout << node->data << " ";
        _print(node->right);
    }
    Node* _insert(Node *node, int data) {
        if (node == NULL) {
            node = new Node(data);
            return node;
        }
        if (data < node->data) 
            node->left = _insert(node->left, data);
        else if (data > node->data)
            node->right = _insert(node->right, data);
        return node;
    }
    int _depth(Node *node) {
        if (node == NULL)
            return 0;
        return 1 + max(_depth(node->left), _depth(node->right));
    }
    public:
    BST(){
        root = NULL;
    }
    void insert(int x){
        root = _insert(this->root, x);
    }
    void depth(){
        cout << _depth(this->root) << endl;
    }
    
};

int main(){
    BST *bst = new BST();

    int n;
    while(cin >> n){
        if(n == 0)
            break;
        else 
            bst->insert(n);
    }
    bst->depth();
    return 0;
}
